src/engine/strat_engine/filesystem.rs

```diff
@@ -47,11 +54,25 @@ impl StratFilesystem {
         }
     }
 
-    pub fn check(&self, dm: &DM) -> EngineResult<FilesystemStatus> {
+    /// check if filesystem is getting full and needs to be extended
+    /// TODO: deal with the thindev in a Fail state.
+    pub fn check(&mut self, dm: &DM) -> EngineResult<FilesystemStatus> {
         match try!(self.thin_dev.status(dm)) {
-            ThinStatus::Good((_mapped, _highest)) => {
-                // TODO: check if filesystem is getting full and might need to
-                // be extended (hint: use statfs(2))
+            ThinStatus::Good((_, _)) => {
+                let mount_point = try!(self.get_mount_point());
+                let (fs_total_bytes, fs_total_used_bytes) = try!(fs_usage(&mount_point));
```

Comment:
If ```fs_usage``` returns total bytes and total free bytes then it is unnecessary to calculate free bytes here. Please try that.